actions/views/channel_sidebar.ts

```diff
@@ -93,10 +97,74 @@ export function adjustTargetIndexForMove(state: GlobalState, categoryId: string,
     // the channel being removed. For example, if we're moving channelA from [channelA, channelB, channelC] to
     // [channelB, channelA, channelC], newIndex would currently be 2 (which comes after channelB), but we need
     // it to be 1 (which comes after channelB once channelA is removed).
-    const sourceIndex = category.channel_ids.indexOf(channelId);
+    const sourceIndex = category.channel_ids.indexOf(channelIds[0]);
     if (sourceIndex !== -1 && sourceIndex < newIndex) {
         newIndex -= 1;
     }
 
-    return newIndex;
+    return Math.max(newIndex - removedChannelsAboveInsert.length, 0);
+}
+
+export function clearChannelSelection() {
+    return (dispatch: DispatchFunc) => {
+        return dispatch({
+            type: ActionTypes.MULTISELECT_CHANNEL_CLEAR,
+        });
+    };
+}
+
+export function multiSelectChannel(channelId: string) {
+    return {
+        type: ActionTypes.MULTISELECT_CHANNEL,
+        data: channelId,
+    };
+}
+
+export function multiSelectChannelAdd(channelId: string) {
+    return {
+        type: ActionTypes.MULTISELECT_CHANNEL_ADD,
+        data: channelId,
+    };
+}
+
+export function multiSelectChannelTo(channelId: string) {
+    return (dispatch: DispatchFunc, getState: () => GlobalState) => {
+        const state: GlobalState = getState();
+        const selectedChannelIds = state.views.channelSidebar.selectedChannelIds;
+        const lastSelected = state.views.channelSidebar.lastSelectedChannel;
+
+        // Nothing already selected
+        if (!selectedChannelIds.length) {
+            return dispatch({
+                type: ActionTypes.MULTISELECT_CHANNEL,
+                data: channelId,
+            });
+        }
+
+        const allChannelsIdsInOrder = getChannelsInCategoryOrder(state).map((channel) => channel.id);
+        const indexOfNew: number = allChannelsIdsInOrder.indexOf(channelId);
```

Comment:
Could we check `channelId === lastSelected` here instead?

Reply:
We're getting the index here for selection, I don't think that will help :P